File: ps/string-to-integer-atoi.cpp
class Solution {
public:
    int isValidChar(char chr) {
        if(chr == '+' || chr == '-' || ('0' <= chr && chr <= '9')) {
            return true;
        }
        return false;
    }
    int myAtoi(string str) {
        int i;
        long res = 0;
        bool isPositive = true;
        for(i=0;i<str.size();i++) {
            if(str[i] != ' ') break;
        }
        if (!isValidChar(str[i])) return 0;
        if(str[i] == '-') { isPositive = false; i++; }
        else if(str[i] == '+') { i++; }
        
        for(;i<str.size();i++) {
            if(!isValidChar(str[i]) || str[i] =='+' || str[i] == '-') break;
            if(isPositive) res = res*10 + (str[i] - '0');
            else res = res*10 - (str[i] - '0');
            if (res > INT_MAX) return INT_MAX;
            if (res < INT_MIN) return INT_MIN;
        }
        return res;
    }
};

File: ps/string-to-integer-atoi.py
class Solution:
    def myAtoi(self, s: str) -> int:
        sign, i = 1, 0
        while i < len(s) and s[i] == " ": i += 1

        if i < len(s) and s[i] in ['-', '+']:
            if s[i] == '-':
                sign = -1
            i += 1

        num = 0
        while i < len(s) and s[i].isdigit():
            num = num * 10 + int(s[i])
            i += 1
            if num > 2 ** 31: num = 2 ** 31

        num *= sign
        if num == 2 ** 31: num -= 1

        return num
